### PR TITLE
Refactor db and collection to prevent multiple rounds of options validation

### PR DESCRIPTION
**Pull request by kvwalker**
Fixes NODE-1768

### CODE REVIEW

File: lib/collection.js [hunk 1]
```diff
@@ -535,6 +454,7 @@ const insertManySchema = {
   wtimeout: { type: 'number' },
   j: { type: 'boolean' },
   serializeFunctions: { type: 'boolean' },
+  forceServerObjectId: { type: 'boolean' },
```
Comment (mbroadst) [inline lib/collection.js]: `default: false` according to the documentation above

File: lib/collection.js [hunk 2]
```diff
@@ -2046,9 +1944,10 @@ const findOneAndUpdateSchema = {
   update: { overrideOnly: true },
   new: { overrideOnly: true },
   remove: { overrideOnly: true },
-  serializeFunctions: { type: 'function' },
+  serializeFunctions: { type: 'boolean' },
   checkKeys: { overrideOnly: true },
-  readPreference: { overrideOnly: true }
+  readPreference: { overrideOnly: true },
+  fsync: { type: 'number' }
```
Comment (mbroadst) [inline lib/collection.js]: fsync is deprecated, but should we document it anyway if we're already going to add it in the schema?

File: lib/collection.js [hunk 3]
```diff
@@ -2122,8 +2021,9 @@ const findAndModifySchema = {
   session: { type: ClientSession },
   arrayFilters: { type: 'array' },
   readPreference: { overrideOnly: true },
-  serializeFunctions: { type: 'function' },
-  checkKeys: { overrideOnly: true }
+  serializeFunctions: { type: 'boolean' },
+  checkKeys: { overrideOnly: true },
+  collation: { type: 'object' }
```
Comment (mbroadst) [inline lib/collection.js]: this is not documented above

File: lib/collection.js [hunk 4]
```diff
@@ -2241,7 +2141,10 @@ const aggregateSchema = {
   session: { type: ClientSession },
   promiseLibrary: { type: 'function', overrideOnly: true },
   cursorFactory: { overrideOnly: true },
-  optionsValidationLevel: { overrideOnly: true }
+  optionsValidationLevel: { overrideOnly: true },
+  w: { type: ['number', 'string'] },
```
Comment (mbroadst) [inline lib/collection.js]: these three are not documented above

File: lib/mongo_client.js [hunk 5]
```diff
@@ -440,7 +441,8 @@ MongoClient.connect = function(url, options, callback) {
  * @return {ClientSession} the newly established session
  */
 const startSessionSchema = {
-  explicit: { overrideOnly: true }
+  explicit: { overrideOnly: true },
+  causalConsistency: { type: 'boolean' }
```
Comment (mbroadst) [inline lib/mongo_client.js]: document above

File: lib/operations/collection_ops.js [hunk 6]
```diff
@@ -55,6 +61,118 @@ const toError = require('../utils').toError;
...
+/**
+ * Perform an aggregate operation. See Collection.prototype.aggregate for more information.
+ *
+ * @method
+ * @param {Collection} a Collection instance.
+ * @param {object} [pipeline=[]] Array containing all the aggregation framework commands for the execution.
+ * @param {object} [options] Optional settings. See Collection.prototype.aggregate for a list of options.
+ * @param {Collection~aggregationCallback} callback The command result callback
+ */
+const aggregateSchema = {
```
Comment (mbroadst) [inline lib/operations/collection_ops.js]: is this just the same schema defined above in `collection`? Can you give a bit of detail here about the need to duplicate, and why we have `applyDefaults`?

File: lib/operations/collection_ops.js [hunk 7]
```diff
@@ -68,8 +186,8 @@ function bulkWrite(coll, operations, options, callback) {
   // Create the bulk operation
   const bulk =
     options.ordered === true
-      ? coll.initializeOrderedBulkOp(options)
-      : coll.initializeUnorderedBulkOp(options);
+      ? ordered(coll.s.topology, coll, options)
```
Comment (mbroadst) [inline lib/operations/collection_ops.js]: I realize this isn't the first time we have called them this, but it occurs to me now that importing these functions as `unordered`/`ordered` is _way_ too generic. Can we call the operations `initializeOrderedOp(eration)` and the same for unordered?

File: lib/operations/collection_ops.js [hunk 8]
```diff
@@ -444,6 +571,171 @@ function ensureIndex(coll, fieldOrSpec, options, callback) {
...
+/**
+ * Creates a cursor for a query that can be used to iterate over results from MongoDB
+ *
+ * @method
+ * @param {object} [query={}] The cursor query object.
+ * @param {object} [options] Optional settings. See Collection.prototype.find for a list of options.
+ * @throws {MongoError}
+ * @return {Cursor}
+ */
+const findSchema = {
```
Comment (mbroadst) [inline lib/operations/collection_ops.js]: again I think this doesn't quite belong here, and duplicates the existing schema. At very least we should move the schemas somewhere we can reuse them, but also I'm not quite sure I understand why we need to reapply this here.

**Comment by mbroadst** [comment]
@kvwalker I slowed down on review because I'd like to further discuss `applyDefaults` and the need for schema duplication before we move forward

**Comment by kvwalker** [comment]
We can definitely talk about this more in-depth, but to give some context, `applyDefaults` is necessary for certain operations (like `find`) that are called by operations. The goal is to only validate options once, but skipping options validations results in skipping the application of defaults. A good example is `countDocuments`, which calls `collection.aggregate`. If we do not include `applyDefaults` or options validate, the defaults in the `aggregateSchema` (like `cursor = {}`) will not be applied and an error will occur.

**Comment by kvwalker** [comment]
Closing this to separate out `applyDefaults` work in [NODE-1805](https://jira.mongodb.org/browse/NODE-1805)

**Pull request closed without merging**